### PR TITLE
Union find

### PR DESCRIPTION
**Pull request by eggrobin**
One Ring to find them, One Ring to bring them all and in the darkness bind them.

### CODE REVIEW

File: base/disjoint_sets.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  SubsetProperties<T> const& properties();
+
+ private:
+  explicit Subset(not_null<SubsetNode<T>*> node);
+
+  not_null<SubsetNode<T>*> const node_;
+
+  friend Subset<T> Unite<>(Subset<T> left, Subset<T> right);
+  friend bool operator==(Subset left, Subset right) {
+    return left.node_ == right.node_;
```
Comment (pleroy) [inline base/disjoint_sets.hpp]: Put this in the public part.

File: base/disjoint_sets.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+template<typename T>
+class SubsetProperties {
+ public:
+  void MergeWith(SubsetProperties& other) {}
+};
+
+// The arguments are invalidated; the result may be used to get information
+// about the united subset.
+template<typename T>
+Subset<T> Unite(Subset<T> left, Subset<T> right);
```
Comment (pleroy) [inline base/disjoint_sets.hpp]: Make this a static member of `Subset` just like `Find`.

File: base/disjoint_sets.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+template<typename T>
+class SubsetNode;
```
Comment (pleroy) [inline base/disjoint_sets.hpp]: Make this `Subset::Node`.

File: base/disjoint_sets.hpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <list>
+
+#include "base/not_null.hpp"
+
+namespace principia {
+namespace base {
```
Comment (pleroy) [inline base/disjoint_sets.hpp]: Should this have an `internal_disjoint_sets` namespace?

Comment (eggrobin) [inline base/disjoint_sets.hpp]: No need, we're not `using` anything.

File: base/disjoint_sets.hpp (hunk 1 above)
Comment (eggrobin) [inline base/disjoint_sets.hpp]: Done.

File: base/disjoint_sets.hpp (hunk 3 above)
Comment (eggrobin) [inline base/disjoint_sets.hpp]: Done, also Properties, and Node::Get.

File: base/disjoint_sets.hpp (hunk 2 above)
Comment (eggrobin) [inline base/disjoint_sets.hpp]: Done.

**Pull request merged**